src/interpreter/terminator/mod.rs

```diff
@@ -472,23 +475,36 @@ impl<'a, 'tcx> EvalContext<'a, 'tcx> {
 
                         // Interpreter magic: insert an intermediate pointer, so we can skip the
                         // intermediate function call.
-                        // FIXME: this is a memory leak, should probably add the pointer to the
-                        // current stack.
-                        let first = self.value_to_ptr_dont_use(args[0].0, args[0].1)?;
-                        args[0].0 = Value::ByVal(PrimVal::from_ptr(first));
+                        let ptr = match args[0].0 {
```

Comment:
This code makes every element of `temporaries` a `Value::ByRef`. Can we just make `temporaries` into a `Vec<Pointer>`?

The way I see it, no accesses go through `temporaries` and they're only used for cleanup so it makes sense to be a list of `Pointer`s. (Heck, even just a list of allocation ids should do it.)